pkg/route/secret/manager.go

```diff
@@ -0,0 +1,146 @@
+package secret
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	routev1 "github.com/openshift/api/route/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type Manager struct {
+	monitor SecretMonitor
+	// map of handlerRegistrations
+	registeredHandlers map[ObjectKey]SecretEventHandlerRegistration
+
+	lock sync.RWMutex
+
+	// monitors are the producer of the resourceChanges queue
+	resourceChanges workqueue.RateLimitingInterface
+
+	secretHandler cache.ResourceEventHandlerFuncs
+}
+
+func NewManager(kubeClient *kubernetes.Clientset, queue workqueue.RateLimitingInterface) *Manager {
+	return &Manager{
+		monitor:            NewSecretMonitor(kubeClient),
+		lock:               sync.RWMutex{},
+		resourceChanges:    queue,
+		registeredHandlers: make(map[ObjectKey]SecretEventHandlerRegistration),
+
+		// default secret handler
+		secretHandler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) {},
+			UpdateFunc: func(oldObj, newObj interface{}) {},
+			DeleteFunc: func(obj interface{}) {},
+		},
+	}
+}
+
+func (m *Manager) WithSecretHandler(handler cache.ResourceEventHandlerFuncs) *Manager {
+	m.secretHandler = handler
+	return m
+}
+
+func (m *Manager) RegisterRoute(parent *routev1.Route) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	var secretName string
+
+	if parent.Spec.TLS.ExternalCertificate != nil && len(parent.Spec.TLS.ExternalCertificate.Name) > 0 {
+		secretName = parent.Spec.TLS.ExternalCertificate.Name
+	} else {
+		return apierrors.NewInternalError(fmt.Errorf("unable to get secret name from route %v", parent))
```

Comment:
why error versus just return nil after doing nothign?

Reply:
Previously, the intension was that `route` should have a referenced secret and return error if it's not.
With current changes, I'm taking `secretName` directly as a function parameter, so this line is not required.